### PR TITLE
Throttler: return app name in check result, synthesize "why throttled" explanation from result

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Followup to https://github.com/vitessio/vitess/pull/15988, this PR further enhances `CheckThrottlerResponse` by adding the name of throttled/granted app.

Why is this needed? Isn't the app name the same as the app in `CheckThrottlerRequest`?

Not necessarily!

- The requesting app might be `vcopier:d666bbfc_169e_11ef_b0b3_0a43f95f28a3:vreplication:online-ddl`. Suppose the request is throttled. Why is it being throttled? There could be many reasons and based on either of the specific identifiers. For example, someone might have issued a `alter vitess_migration throttle all`, in which case the throttled app will be `online-ddl`. So this is what we indicate back to the user.
- There may not be any specific rule for the given app, but it's possible that the `all` app is being throttled. Either because someone actually invoked `UpdateThrottlerConfig --throttle-app=all ...`, or just because `all` is assigned specific metrics, one of which crossed its threshold. This way or another, we want to tell the user the `"all"` app is at "fault" here.

### Check result summary

Based on that, we provide a concise summary, something like `app 'online-ddl' is throttled because 'lag' exceeded its threshold of `5``, or `vreplication is denied access until ...`.

Example actual summaries:
```
online-ddl is granted access
online-ddl is explicitly denied access
online-ddl is denied access due to threads_running metric value 4 exceeding threshold 2
online-ddl is denied access due to lag metric value 94.821447 exceeding threshold 5
online-ddl is denied access due to lag metric value 22.951372 exceeding threshold 5
all is explicitly denied access
```

This is in turn injected in `_vt.vreplication` and then in `_vt.schema_migrations` as a human readable assist for analyzing production situations.

### `reason_throttled` in `_vt.vreplication`

@rohit-nayak-ps you will like this.

Looks like this:
```sql
                   id: 3
             workflow: d741953d_4503_11ef_8761_0a43f95f28a3
               source: keyspace:"commerce" shard:"0" filter:{rules:{match:"_vt_vrp_d741953d450311ef87610a43f95f28a3_20240718124708_" filter:"select `customer_id`
 as `customer_id`, `email` as `email` from `customer`" source_unique_key_columns:"customer_id" target_unique_key_columns:"customer_id" source_unique_key_target_c
olumns:"customer_id" force_unique_key:"PRIMARY"}}
                  pos: MySQL56/414befe7-4358-11ef-bc56-0a43f95f28a3:1-650
             stop_pos: NULL
              max_tps: 9223372036854775807
  max_replication_lag: 9223372036854775807
                 cell:
         tablet_types: in_order:REPLICA,PRIMARY
         time_updated: 1721306835
transaction_timestamp: 1721306834
                state: Stopped
              message: stopped for online DDL cutover
              db_name: vt_commerce
          rows_copied: 0
                 tags:
       time_heartbeat: 1721306835
        workflow_type: 5
       time_throttled: 1721306829
  component_throttled: vcopier
     reason_throttled: online-ddl is denied access due to shard/lag metric value 8.725835 exceeding threshold 5
    workflow_sub_type: 0
 defer_secondary_keys: 0
              options: {}
```

The same is copied to `_vt.schema_migrations`.

## Related Issue(s)

- https://github.com/vitessio/vitess/pull/15988
- https://github.com/vitessio/vitess/issues/15624
- 
## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Review by rohit-nayak-ps — APPROVED** [review]
This is brilliant! 

I will enhance the throttler reporting added to vtadmin in https://github.com/vitessio/vitess/pull/16308 to display the `reason` once this is merged.

File: go/vt/vttablet/tabletserver/throttle/throttler_test.go [hunk 1]
```diff
@@ -803,38 +840,99 @@ func TestIsAppThrottled(t *testing.T) {
...
+			throttled, app := throttler.IsAppThrottled("app_other")
+			assert.True(t, throttled) // falls under "all"
+			assert.Equal(t, "all", app)
+		}
 
 		// continuing previous test, we had 3 throttled apps. "all" is a new app being throttled.
 		assert.Equal(t, 4, throttler.throttledApps.ItemCount())
 	})
-	//
+	// 	//
```
Comment (mattlord) [inline go/vt/vttablet/tabletserver/throttle/throttler_test.go:909]: Looks like we added an extraneous comment indicator.

**Review by mattlord — APPROVED** [review]
Thank you! This is a nice enhancement. ❤️

File: go/vt/vttablet/tabletserver/throttle/check_result.go [hunk 2]
```diff
@@ -85,14 +91,32 @@ func (c *CheckResult) IsOK() bool {
 	return c.StatusCode == http.StatusOK
 }
 
+// Summary returns a human-readable summary of the check result
+func (c *CheckResult) Summary() string {
+	switch c.StatusCode {
+	case http.StatusOK:
+		return fmt.Sprintf("%v is granted access", c.AppName)
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/check_result.go:98]: done

File: go/vt/vttablet/tabletserver/throttle/client.go [hunk 3]
```diff
@@ -31,8 +31,11 @@ const (
 	throttleCheckDuration = 250 * time.Millisecond
 )
 
-var throttleTicks int64
-var throttleInit sync.Once
+var (
+	throttleTicks    int64
+	throttleInit     sync.Once
+	emptyCheckResult = &CheckResult{}
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/client.go:37]: `golang` does nto permit this as a `const`

File: go/vt/vttablet/tabletserver/throttle/throttler_test.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/throttler_test.go:909]: removed

File: go/vt/vttablet/tabletserver/throttle/check_result.go [hunk 4]
```diff
@@ -85,14 +91,32 @@ func (c *CheckResult) IsOK() bool {
 	return c.StatusCode == http.StatusOK
 }
 
+// Summary returns a human-readable summary of the check result
+func (c *CheckResult) Summary() string {
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/check_result.go:95]: I'm ambivalent. There can only be one `String()` function, and I'm not sure what the expected implementation should be. `Summary()` does not capture all of the `CheckResult`'s fields. Should `String()` capture all fields? Should `String()` be human readable or machine readable?

File: go/vt/vttablet/tabletserver/throttle/throttler_test.go [hunk 5]
```diff
@@ -418,6 +418,7 @@ func TestApplyThrottlerConfigMetricThresholds(t *testing.T) {
 			assert.EqualValues(t, 0.3, checkResult.Value) // self lag value
 			assert.EqualValues(t, http.StatusOK, checkResult.StatusCode)
 			assert.Len(t, checkResult.Metrics, 1)
+			assert.Contains(t, checkResult.Summary(), "test is granted access")
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/throttler_test.go]: Fixed.

File: go/vt/vttablet/tabletserver/throttle/check_result.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/check_result.go:95]: Gonna keep this with `Summary()`. We can always easily change it to `String()` later.

**Pull request merged**